node/core/pvf/src/prepare/worker.rs

```diff
@@ -27,11 +27,23 @@ use async_std::{
 	io,
 	os::unix::net::UnixStream,
 	path::{Path, PathBuf},
+	sync::Mutex,
+	task,
 };
+use cpu_time::ProcessTime;
 use parity_scale_codec::{Decode, Encode};
 use sp_core::hexdisplay::HexDisplay;
 use std::{panic, sync::Arc, time::Duration};
 
+/// A multiple of the preparation timeout (CPU time) for which we are willing to wait on the host
+/// in wall clock time. This is lenient because the wall clock can slow down up to a factor of 4
+/// (tested on 32 threads of work on an 8-core machine).
+const PREPARATION_TIMEOUT_WALL_CLOCK_FACTOR: u32 = 4;
+
+/// Some allowed overhead that we account for in the "CPU time monitor" thread's sleeps, on the
+/// child process.
+const PREPARATION_TIMEOUT_OVERHEAD: Duration = Duration::from_millis(50);
```

Comment:
I believe sleeping threads shouldn't use any CPU. It's more for thread overhead, delays in waking up (sleeps are not totally exact), etc. 

Also, this serves the double function of preventing waking up the thread for short waits. Say there is 10ms remaining, we might wait 10ms and wake it up only to find that cpu time went slower than expected and there is still 5ms. We repeat and wait 5ms, then 2ms, then 1ms. To prevent this situation we can add a bit of buffer to the sleep time.

Reply:
Ok, I guess that makes sense. Although why do we add it [here](https://github.com/paritytech/polkadot/blob/e34825aa290d2b05f51f62c3afc6f4c7f95f4bcc/node/core/pvf/src/prepare/worker.rs#L129)?

Reply:
Oh, I figured that if the job completed before the process actually terminates, it's fine, even if it took a bit longer than the timeout. But now I think we should still enforce the timeout there, where we have the exact CPU time the job took.

I still think we should always add this buffer in the child. I ran some more tests and CPU clock always seems at least slightly slower than wall clock. So if we just slept for 5 s wall clock time, only 4.99 s CPU time may have passed.